From e1658dcff0bbf1e0d75c673d0d3b23a329cec564 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Tue, 11 Feb 2025 11:33:50 +0800
Subject: [PATCH 108/137] mtk: mac80211: relax the DFS restrictions for
 scanning

If user turn on the dfs relax flag by entering the following command:
echo 1 > /sys/kernel/debug/ieee80211/phy0/scan_dfs_relax
then
1. Allow the AP to scan while operating on a DFS channel
2. Allow the MLO AP to scan the other bands when 5G band is performing
   CAC (5G band will be skipped).

This patch is rebased from 04af7e0d15d5c2c9b8e3adf40529342d06850b91.

Link: https://github.com/mediatek/mtk-openwrt-feeds/blob/a15454c888f4f4144c50e33b5feef4f247c5f78b/autobuild/unified/filogic/mac80211/25.12/files/package/kernel/mac80211/patches/subsys/0108-mtk-mac80211-relax-the-DFS-restrictions-for-scanning.patch
Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
[plcharriere: relax only scans: with the flag set, a scan is still refused
while CAC runs on the same or an unknown radio, and remain-on-channel
keeps the restriction; access the flag with READ_ONCE()/WRITE_ONCE()]
Assisted-by: Codex:gpt-5.6-sol
Signed-off-by: Paul-Louis Charrière <plcharriere@gmail.com>
---
 include/net/cfg80211.h | 10 ++++++++-
 net/mac80211/scan.c    |  7 ++++---
 net/wireless/debugfs.c | 57 ++++++++++++++++++++++++++++++++++++++++++++++++++
 net/wireless/nl80211.c | 47 +++++++++++++++++++++++++++++++++--------
 net/wireless/scan.c    |  4 ++--
 net/wireless/util.c    | 13 ++++++++-----
 6 files changed, 118 insertions(+), 20 deletions(-)

--- a/include/net/cfg80211.h
+++ b/include/net/cfg80211.h
@@ -6520,6 +6520,8 @@ struct wiphy_nan_capa {
  *
  * @radio: radios belonging to this wiphy
  * @n_radio: number of radios
+ *
+ * @dfs_relax: a flag to relax the DFS restrictions during scanning
  */
 struct wiphy {
 	struct mutex mtx;
@@ -6677,6 +6679,8 @@ struct wiphy {
 	int n_radio;
 	const struct wiphy_radio *radio;
 
+	bool dfs_relax;
+
 	char priv[] __aligned(NETDEV_ALIGN);
 };
 
@@ -7521,11 +7525,15 @@ bool cfg80211_radio_chandef_valid(const
  *
  * @wdev: the wireless device
  * @chan: channel to check
+ * @radio_mask: check the channel under a user-specified radio mask.
+ *	If the radio_mask is 0, then wdev->radio_mask is used
+ *	to check the channel.
  *
  * Return: whether or not the wdev may use the channel
  */
 bool cfg80211_wdev_channel_allowed(struct wireless_dev *wdev,
-				   struct ieee80211_channel *chan);
+				   struct ieee80211_channel *chan,
+				   u32 radio_mask);
 
 /**
  * ieee80211_get_response_rate - get basic rate for a given rate
--- a/net/mac80211/scan.c
+++ b/net/mac80211/scan.c
@@ -607,7 +607,8 @@ static bool __ieee80211_can_leave_ch(str
 	if (!ieee80211_is_radar_required(local, req))
 		return true;
 
-	if (!regulatory_pre_cac_allowed(local->hw.wiphy))
+	if (!regulatory_pre_cac_allowed(local->hw.wiphy) &&
+	    !READ_ONCE(local->hw.wiphy->dfs_relax))
 		return false;
 
 	list_for_each_entry(sdata_iter, &local->interfaces, list) {
@@ -1234,7 +1235,7 @@ int ieee80211_request_ibss_scan(struct i
 				if (tmp_ch->flags & (IEEE80211_CHAN_NO_IR |
 						     IEEE80211_CHAN_DISABLED) ||
 				    !cfg80211_wdev_channel_allowed(&sdata->wdev,
-								   tmp_ch))
+								   tmp_ch, 0))
 					continue;
 
 				local->int_scan_req->channels[n_ch] = tmp_ch;
@@ -1251,7 +1252,7 @@ int ieee80211_request_ibss_scan(struct i
 			if (channels[i]->flags & (IEEE80211_CHAN_NO_IR |
 						  IEEE80211_CHAN_DISABLED) ||
 			    !cfg80211_wdev_channel_allowed(&sdata->wdev,
-							   channels[i]))
+							   channels[i], 0))
 				continue;
 
 			local->int_scan_req->channels[n_ch] = channels[i];
--- a/net/wireless/debugfs.c
+++ b/net/wireless/debugfs.c
@@ -118,6 +118,61 @@ static const struct file_operations ht40
 	.llseek = default_llseek,
 };
 
+static ssize_t scan_dfs_relax_write(struct file *file,
+				    const char __user *user_buf,
+				    size_t count, loff_t *ppos)
+{
+	struct wiphy *wiphy = file->private_data;
+	bool dfs_relax;
+	char buf[16];
+
+	if (count >= sizeof(buf))
+		return -EINVAL;
+
+	if (copy_from_user(buf, user_buf, count))
+		return -EFAULT;
+
+	if (count && buf[count - 1] == '\n')
+		buf[count - 1] = '\0';
+	else
+		buf[count] = '\0';
+
+	if (kstrtobool(buf, &dfs_relax))
+		return -EINVAL;
+
+	WRITE_ONCE(wiphy->dfs_relax, dfs_relax);
+
+	return count;
+}
+
+static ssize_t scan_dfs_relax_read(struct file *file, char __user *user_buf,
+				   size_t count, loff_t *ppos)
+{
+	struct wiphy *wiphy = file->private_data;
+	unsigned int r, offset, buf_size = PAGE_SIZE;
+	char *buf;
+
+	buf = kzalloc(buf_size, GFP_KERNEL);
+	if (!buf)
+		return -ENOMEM;
+
+	offset = scnprintf(buf, buf_size, "dfs relax: %u\n",
+			   READ_ONCE(wiphy->dfs_relax));
+
+	r = simple_read_from_buffer(user_buf, count, ppos, buf, offset);
+
+	kfree(buf);
+
+	return r;
+}
+
+static const struct file_operations scan_dfs_relax_ops = {
+	.write = scan_dfs_relax_write,
+	.read = scan_dfs_relax_read,
+	.open = simple_open,
+	.llseek = default_llseek,
+};
+
 #define DEBUGFS_ADD(name)						\
 	debugfs_create_file(#name, 0444, phyd, &rdev->wiphy, &name## _ops)
 
@@ -138,6 +193,8 @@ void cfg80211_debugfs_rdev_add(struct cf
 	DEBUGFS_ADD(long_retry_limit);
 	DEBUGFS_ADD(ht40allow_map);
 
+	debugfs_create_file("scan_dfs_relax", 0644, phyd, &rdev->wiphy,
+			    &scan_dfs_relax_ops);
 	for (i = 0; i < rdev->wiphy.n_radio; i++) {
 		radiod = rdev->wiphy.radio_cfg[i].radio_debugfsdir;
 		DEBUGFS_RADIO_ADD(radio_rts_threshold, i);
--- a/net/wireless/nl80211.c
+++ b/net/wireless/nl80211.c
@@ -11119,7 +11119,8 @@ int nl80211_parse_random_mac(struct nlat
 }
 
 static bool cfg80211_off_channel_oper_allowed(struct wireless_dev *wdev,
-					      struct ieee80211_channel *chan)
+					      struct ieee80211_channel *chan,
+					      bool dfs_relax)
 {
 	unsigned int link_id;
 	bool all_ok = true;
@@ -11127,13 +11128,36 @@ static bool cfg80211_off_channel_oper_al
 
 	lockdep_assert_wiphy(wdev->wiphy);
 
-	if (!cfg80211_wdev_channel_allowed(wdev, chan))
+	if (!cfg80211_wdev_channel_allowed(wdev, chan, 0))
 		return false;
 
-	if (!cfg80211_beaconing_iface_active(wdev))
+	radio_idx = cfg80211_get_radio_idx_by_chan(wdev->wiphy, chan);
+
+	if (dfs_relax) {
+		for_each_valid_link(wdev, link_id) {
+			struct cfg80211_chan_def *chandef;
+			int link_radio_idx;
+
+			if (!wdev->links[link_id].cac_started)
+				continue;
+
+			chandef = wdev_chandef(wdev, link_id);
+			if (!chandef || !chandef->chan)
+				return false;
+
+			link_radio_idx =
+				cfg80211_get_radio_idx_by_chan(wdev->wiphy,
+							       chandef->chan);
+			if (radio_idx < 0 || link_radio_idx < 0 ||
+			    link_radio_idx == radio_idx)
+				return false;
+		}
+
 		return true;
+	}
 
-	radio_idx = cfg80211_get_radio_idx_by_chan(wdev->wiphy, chan);
+	if (!cfg80211_beaconing_iface_active(wdev))
+		return true;
 
 	/*
 	 * FIXME: check if we have a free radio/link for chan
@@ -11356,7 +11380,7 @@ static int nl80211_trigger_scan(struct s
 			/* Ignore disabled / no primary channels */
 			if (chan->flags & IEEE80211_CHAN_DISABLED ||
 			    chan->flags & IEEE80211_CHAN_S1G_NO_PRIMARY ||
-			    !cfg80211_wdev_channel_allowed(wdev, chan))
+			    !cfg80211_wdev_channel_allowed(wdev, chan, 0))
 				continue;
 
 			request->req.channels[i] = chan;
@@ -11364,6 +11388,11 @@ static int nl80211_trigger_scan(struct s
 		}
 	} else {
 		enum nl80211_band band;
+		u32 radio_mask;
+
+		err = nl80211_parse_vif_radio_mask(info, &radio_mask);
+		if (err < 0)
+			goto out_free;
 
 		/* all channels */
 		for (band = 0; band < NUM_NL80211_BANDS; band++) {
@@ -11379,7 +11408,7 @@ static int nl80211_trigger_scan(struct s
 				if (chan->flags & IEEE80211_CHAN_DISABLED ||
 				    chan->flags &
 					    IEEE80211_CHAN_S1G_NO_PRIMARY ||
-				    !cfg80211_wdev_channel_allowed(wdev, chan))
+				    !cfg80211_wdev_channel_allowed(wdev, chan, radio_mask))
 					continue;
 
 				request->req.channels[i] = chan;
@@ -11399,7 +11428,7 @@ static int nl80211_trigger_scan(struct s
 		struct ieee80211_channel *chan = request->req.channels[i];
 
 		/* if we can go off-channel to the target channel we're good */
-		if (cfg80211_off_channel_oper_allowed(wdev, chan))
+		if (cfg80211_off_channel_oper_allowed(wdev, chan, READ_ONCE(wiphy->dfs_relax)))
 			continue;
 
 		if (!cfg80211_wdev_on_sub_chan(wdev, chan, true)) {
@@ -14614,7 +14643,7 @@ static int nl80211_remain_on_channel(str
 	if (err)
 		return err;
 
-	if (!cfg80211_off_channel_oper_allowed(wdev, chandef.chan)) {
+	if (!cfg80211_off_channel_oper_allowed(wdev, chandef.chan, false)) {
 		const struct cfg80211_chan_def *oper_chandef, *compat_chandef;
 
 		oper_chandef = wdev_chandef(wdev, link_id);
@@ -14848,7 +14877,7 @@ static int nl80211_tx_mgmt(struct sk_buf
 		return -EINVAL;
 
 	if (params.offchan &&
-	    !cfg80211_off_channel_oper_allowed(wdev, chandef.chan))
+	    !cfg80211_off_channel_oper_allowed(wdev, chandef.chan, false))
 		return -EBUSY;
 
 	params.link_id = nl80211_link_id_or_invalid(info->attrs);
--- a/net/wireless/scan.c
+++ b/net/wireless/scan.c
@@ -997,7 +997,7 @@ static int cfg80211_scan_6ghz(struct cfg
 			ieee80211_get_channel(&rdev->wiphy, ap->center_freq);
 
 		if (!chan || chan->flags & IEEE80211_CHAN_DISABLED ||
-		    !cfg80211_wdev_channel_allowed(rdev_req->req.wdev, chan))
+		    !cfg80211_wdev_channel_allowed(rdev_req->req.wdev, chan, 0))
 			continue;
 
 		for (i = 0; i < rdev_req->req.n_channels; i++) {
@@ -3575,7 +3575,7 @@ int cfg80211_wext_siwscan(struct net_dev
 			/* ignore disabled channels */
 			chan = &wiphy->bands[band]->channels[j];
 			if (chan->flags & IEEE80211_CHAN_DISABLED ||
-			    !cfg80211_wdev_channel_allowed(creq->req.wdev, chan))
+			    !cfg80211_wdev_channel_allowed(creq->req.wdev, chan, 0))
 				continue;
 
 			/* If we have a wireless request structure and the
--- a/net/wireless/util.c
+++ b/net/wireless/util.c
@@ -3056,21 +3056,24 @@ bool cfg80211_radio_chandef_valid(const
 EXPORT_SYMBOL(cfg80211_radio_chandef_valid);
 
 bool cfg80211_wdev_channel_allowed(struct wireless_dev *wdev,
-				   struct ieee80211_channel *chan)
+				   struct ieee80211_channel *chan,
+				   u32 radio_mask)
 {
 	struct wiphy *wiphy = wdev->wiphy;
 	const struct wiphy_radio *radio;
 	struct cfg80211_chan_def chandef;
-	u32 radio_mask;
+	u32 mask = radio_mask & wdev->radio_mask;
 	int i;
 
-	radio_mask = wdev->radio_mask;
-	if (!wiphy->n_radio || radio_mask == BIT(wiphy->n_radio) - 1)
+	if (!radio_mask)
+		mask = wdev->radio_mask;
+
+	if (!wiphy->n_radio || mask == BIT(wiphy->n_radio) - 1)
 		return true;
 
 	cfg80211_chandef_create(&chandef, chan, NL80211_CHAN_HT20);
 	for (i = 0; i < wiphy->n_radio; i++) {
-		if (!(radio_mask & BIT(i)))
+		if (!(mask & BIT(i)))
 			continue;
 
 		radio = &wiphy->radio[i];
